keps/sig-scheduling/4816-dra-prioritized-list/README.md

```diff
@@ -0,0 +1,1134 @@
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+A `DeviceRequest` that populates the `FirstAvailableOf` field must *not*
+populate the `DeviceClassName` field. The `required` validation on this field
+will be relaxed. This allows existing clients to differentiate between claims
+they understand (with `DeviceClassName`) and those they do not (without
+`DeviceClassName` but with the new field). Clients written for 1.31, when
+`DeviceClassName` was required, were requested to include this logic, and the
+in-tree components have been built in this way.
+
+```go
+// DeviceRequest is a request for devices required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for several identical devices.
+type DeviceRequest struct {
+    // Name can be used to reference this request in a pod.spec.containers[].resources.claims
+    // entry and in a constraint of the claim.
+    //
+    // Must be a DNS label.
+    //
+    // +required
+    Name string
+
+    // DeviceClassName references a specific DeviceClass, which can define
+    // additional configuration and selectors to be inherited by this
+    // request.
+    //
+    // Either a class or FirstAvailableOf requests are required in DeviceClaim.Requests.
+    // When this request is part of the FirstAvailableOf list, a class is required. Nested
+    // FirstAvailableOf requests are not allowed
+    //
+    // Which classes are available depends on the cluster.
+    //
+    // Administrators may use this to restrict which devices may get
+    // requested by only installing classes with selectors for permitted
+    // devices. If users are free to request anything without restrictions,
+    // then administrators can create an empty DeviceClass for users
+    // to reference.
+    //
+    // +optional
+    // +oneOf=deviceRequestType
+    DeviceClassName string
+
+    // FirstAvailableOf contains subrequests, exactly one of which must be satisfied
+    // in order to satisfy this request. This field may only be set in the
+    // entries of DeviceClaim.Requests. It must not be set in DeviceRequest
+    // instances that themselves are part of a FirstAvailableOf.
+    //
+    // +optional
+    // +oneOf=deviceRequestType
+    FirstAvailableOf []DeviceRequest
+
+    ...
```

Comment:
The [original API version](https://github.com/kubernetes/enhancements/pull/4871/commits/76a1f298b7943c0d41b2a43cd3deedc0909165b1#diff-cb6e98e5e7618608c36970b6e2ea9ddcca77609834639b4bf6324f413b4f44b4R257) I pushed had `RankedDeviceRequests` that was a sibling to `Requests` in `DeviceClaim`, but @pohly was [concerned](https://github.com/kubernetes/enhancements/pull/4871/commits/76a1f298b7943c0d41b2a43cd3deedc0909165b1#r1778151870) about older clients treating it as an empty claim.

I do think it's cleaner not to have the one "unrolled" one-of along with the embedded list of one-ofs, which we have now and is sort of weird.

Reply:
Given  that this KEP is merged, I will shift focus elsewhere, but I do want to redvisit this before code deadline.  We CAN move things around between alpha and beta, as long as it can round-trip.